File: python/axint/bundle_hash.py
"""
Canonical hash over the bytes that leave the registry at install time.

Mirror of axint/src/core/bundle-hash.ts and the registry Worker's
bundle-hash.ts. Must stay byte-identical with both — anything that
changes serialization here has to ship to the other two in the same
release, or every install will fail bundle verification.

The contract:
  * UTF-8 JSON, keys in alphabetical order, no whitespace
  * Missing optional fields are explicit null, never absent
  * SHA-256, lower-case hex, 64 chars
"""

from __future__ import annotations

import hashlib
import json
from typing import TypedDict


class _SwiftOutput(TypedDict):
    swift_output: str


# swift_output is the only field every bundle must carry (the registry
# uses it as the canonical compiled artifact). The other three are
# optional for the same reason they're optional on the TS side: a
# Python-only or TS-only package legitimately omits the other source.
class BundleContents(_SwiftOutput, total=False):
    ts_source: str | None
    py_source: str | None
    plist_fragment: str | None


BUNDLE_HASH_ALGORITHM = "sha256"
BUNDLE_HASH_HEX_LENGTH = 64


def canonicalize_bundle(bundle: BundleContents) -> str:
    # .get() returns None for missing keys; we use it directly instead of
    # `or None` so empty strings stay as empty strings, matching the TS
    # `??` operator on the other side of the wire.
    normalized = {
        "plist_fragment": bundle.get("plist_fragment"),
        "py_source": bundle.get("py_source"),
        "swift_output": bundle["swift_output"],
        "ts_source": bundle.get("ts_source"),
    }
    return json.dumps(normalized, separators=(",", ":"))


def hash_bundle(bundle: BundleContents) -> str:
    canonical = canonicalize_bundle(bundle)
    return hashlib.sha256(canonical.encode("utf-8")).hexdigest()

File: src/core/bundle-hash.ts
/**
 * Canonical hash over the bytes that leave the registry at install time.
 *
 * The registry stores this hash on publish and returns it on install so
 * `axint add` can verify the bundle wasn't altered between the two
 * endpoints. We canonicalize on the TS side, the server side (Workers
 * runtime), and anywhere else a bundle round-trips — if any of those
 * three disagree on bytes, the hash diverges and we catch it before the
 * files hit the user's disk.
 *
 * The contract:
 *   • UTF-8 JSON, keys in alphabetical order, no whitespace
 *   • Missing optional fields are explicit `null`, never `undefined`
 *   • SHA-256, lower-case hex, 64 chars
 */

export interface BundleContents {
  ts_source?: string | null;
  py_source?: string | null;
  swift_output: string;
  plist_fragment?: string | null;
}

export function canonicalizeBundle(bundle: BundleContents): string {
  const normalized = {
    plist_fragment: bundle.plist_fragment ?? null,
    py_source: bundle.py_source ?? null,
    swift_output: bundle.swift_output,
    ts_source: bundle.ts_source ?? null,
  };
  return JSON.stringify(normalized);
}

export async function hashBundle(bundle: BundleContents): Promise<string> {
  const canonical = canonicalizeBundle(bundle);
  const bytes = new TextEncoder().encode(canonical);
  const digest = await crypto.subtle.digest("SHA-256", bytes);
  return Array.from(new Uint8Array(digest), (b) => b.toString(16).padStart(2, "0")).join(
    ""
  );
}

export const BUNDLE_HASH_ALGORITHM = "sha256" as const;
export const BUNDLE_HASH_HEX_LENGTH = 64;
